app/src/main/java/org/thoughtcrime/securesms/conversation/v2/ConversationActivityV2.kt

```diff
@@ -2090,41 +2215,95 @@ class ConversationActivityV2 : PassphraseRequiredActionBarActivity(), InputBarDe
         endActionMode()
     }
 
+    private fun saveAttachments(message: MmsMessageRecord) {
+        val attachments: List<SaveAttachmentTask.Attachment?> = Stream.of(message.slideDeck.slides)
+            .filter { s: Slide -> s.uri != null && (s.hasImage() || s.hasVideo() || s.hasAudio() || s.hasDocument()) }
+            .map { s: Slide -> SaveAttachmentTask.Attachment(s.uri!!, s.contentType, message.dateReceived, s.fileName.orNull()) }
+            .toList()
+        if (attachments.isNotEmpty()) {
+            val saveTask = SaveAttachmentTask(this)
+            saveTask.executeOnExecutor(AsyncTask.THREAD_POOL_EXECUTOR, *attachments.toTypedArray())
+            if (!message.isOutgoing) { sendMediaSavedNotification() }
+            return
+        }
+        // Implied else that there were no attachment(s)
+        Toast.makeText(this, resources.getString(R.string.attachmentsSaveError), Toast.LENGTH_LONG).show()
+    }
+
+    private fun hasPermission(permission: String): Boolean {
+        val result = ContextCompat.checkSelfPermission(this, permission)
+        return result == PackageManager.PERMISSION_GRANTED
+    }
+
     override fun saveAttachment(messages: Set<MessageRecord>) {
         val message = messages.first() as MmsMessageRecord
 
-        // Do not allow the user to download a file attachment before it has finished downloading
+        // Note: The save option is only added to the menu in ConversationReactionOverlay.getMenuActionItems
+        // if the attachment has finished downloading, so we don't really have to check for message.isMediaPending
+        // here - but we'll do it anyway and bail should that be the case as a defensive programming strategy.
         if (message.isMediaPending) {
-            Toast.makeText(this, resources.getString(R.string.conversation_activity__wait_until_attachment_has_finished_downloading), Toast.LENGTH_LONG).show()
+            Log.w(TAG, "Somehow we were asked to download an attachment before it had finished downloading - aborting download.")
             return
         }
 
+        // On Android versions below 30 we require the WRITE_EXTERNAL_STORAGE permission to save attachments.
+        // However, we would like to  on more recent Android API versions there is scoped storage
+        // If we already have permission to write to external storage then just get on with it & return..
+        //
+        // Android versions will j
```

Comment:
Missing part of the comment on that last sentence?

Reply:
Cleaned up / fixed

Reply:
Still says `// Android versions will j`